compiler/rustc_codegen_ssa/src/codegen_attrs.rs

```diff
@@ -60,6 +61,21 @@ fn codegen_fn_attrs(tcx: TyCtxt<'_>, did: DefId) -> CodegenFnAttrs {
 
     let supported_target_features = tcx.supported_target_features(LOCAL_CRATE);
 
+    // In some cases, attribute are only valid on functions, but it's the `check_attr`
+    // pass that check that they aren't used anywhere else, rather this module.
+    // In these cases, we bail from performing further checks that are only meaningful for
+    // functions (such as calling `fn_sig`, which ICEs if given a non-function). We also
+    // report a delayed bug, just in case `check_attr` isn't doing its job.
+    let validate_fn_only_attr = |attr_sp| -> bool {
+        let def_kind = tcx.def_kind(did);
+        if let DefKind::Fn | DefKind::AssocFn = def_kind {
```

Comment:
DefKind could be Variant or Ctor here too. Why do we exclude them ?

Reply:
I don't think either of the attributes here can be applied to `Variant`s or `Ctor`s, so I left them out. However, adding them won't do any harm, and may be better for forward compatibility, in case this gets applied to other attributes in future. I'll add them.